Adding Anuga

stoiver opened the pull request:
<!--
Thank you very much for putting in this recipe PR!

This repository is very active, so if you need help with a PR, please let the
right people know. There are language-specific teams for reviewing recipes.

| Language        | Name of review team           |
| --------------- | ----------------------------- |
| python          | `@conda-forge/help-python`    |
| python/c hybrid | `@conda-forge/help-python-c`  |
| r               | `@conda-forge/help-r`         |
| java            | `@conda-forge/help-java`      |
| nodejs          | `@conda-forge/help-nodejs`    |
| c/c++           | `@conda-forge/help-c-cpp`     |
| perl            | `@conda-forge/help-perl`      |
| Julia           | `@conda-forge/help-julia`     |
| ruby            | `@conda-forge/help-ruby`      |
| other           | `@conda-forge/staged-recipes` |

Once the PR is ready for review, please mention one of the teams above in a
new comment. i.e. `@conda-forge/help-some-language, ready for review!`
Then, a bot will label the PR as 'review-requested'.

Due to GitHub limitations, first time contributors to conda-forge are unable
to ping conda-forge teams directly, but you can [ask a bot to ping the team][1]
using a special command in a comment on the PR to get the attention of the
`staged-recipes` team. You can also consider asking on our [Gitter channel][2]
if your recipe isn't reviewed promptly.

[1]: https://conda-forge.org/docs/maintainer/infrastructure.html#conda-forge-admin-please-ping-team
[2]: https://gitter.im/conda-forge/conda-forge.github.io
-->



@conda-forge/help-python-c this package needs to use the mingw compiler on windows. I tried to setup the build section of the meta.yaml to use m2w64-toolchain and libpython on windows. Hope it works. 

Checklist
- [X] Title of this PR is meaningful: e.g. "Adding my_nifty_package", not "updated meta.yaml".
- [X] License file is packaged (see [here](https://github.com/conda-forge/staged-recipes/blob/5eddbd7fc9d1502169089da06c3688d9759be978/recipes/example/meta.yaml#L64-L73) for an example).
- [X] Source is from official source.
- [X] Package does not vendor other packages. (If a package uses the source of another package, they should be separate packages or the licenses of all packages need to be packaged).
- [X] If static libraries are linked in, the license of the static library is packaged.
- [X] Build number is 0.
- [X] A tarball (`url`) rather than a repo (e.g. `git_url`) is used in your recipe (see [here](https://conda-forge.org/docs/maintainer/adding_pkgs.html#build-from-tarballs-not-repos) for more details).
- [X] GitHub users listed in the maintainer section have posted a comment confirming they are willing to be listed there.
- [X] When in trouble, please check our [knowledge base documentation](https://conda-forge.org/docs/maintainer/knowledge_base.html) before pinging a team.

conda-forge-linter commented:
Hi! This is the friendly automated conda-forge-linting service.

I wanted to let you know that I linted all conda-recipes in your PR (```recipes/anuga```) and found some lint.

Here's what I've got...


For **recipes/anuga**:

 * Selectors are suggested to take a ``<two spaces>#<one space>[<expression>]`` form. See lines [15]
 * `noarch` packages can't have selectors. If the selectors are necessary, please remove `noarch: python`.
 * noarch: python recipes are required to have a lower bound on the python version. Typically this means putting `python >=3.6` in **both** `host` and `run` but you should check upstream for the package's Python compatibility.

For **recipes/anuga**:

 * Recipes should usually depend on `matplotlib-base` as opposed to `matplotlib` so that runtime environments do not require large packages like `qt`.

conda-forge-linter commented:
Hi! This is the friendly automated conda-forge-linting service.

I wanted to let you know that I linted all conda-recipes in your PR (```recipes/anuga```) and found some lint.

Here's what I've got...


For **recipes/anuga**:

 * Failed to even lint the recipe, probably because of a conda-smithy bug :cry:. This likely indicates a problem in your `meta.yaml`, though. To get a traceback to help figure out what's going on, install conda-smithy and run `conda smithy recipe-lint .` from the recipe directory.

conda-forge-linter commented:
Hi! This is the friendly automated conda-forge-linting service.

I just wanted to let you know that I linted all conda-recipes in your PR (```recipes/anuga```) and found it was in an excellent condition.

conda-forge-linter commented:
Hi! This is the friendly automated conda-forge-linting service.

I just wanted to let you know that I linted all conda-recipes in your PR (```recipes/anuga```) and found it was in an excellent condition.


I do have some suggestions for making it better though...


For **recipes/anuga**:

 * Recipes should usually depend on `matplotlib-base` as opposed to `matplotlib` so that runtime environments do not require large packages like `qt`.

conda-forge-linter commented:
Hi! This is the friendly automated conda-forge-linting service.

I just wanted to let you know that I linted all conda-recipes in your PR (```recipes/anuga```) and found it was in an excellent condition.

stoiver commented:
@conda-forge/help-python-c This recipe should be ready for review.

carterbox commented on recipes/anuga/meta.yaml:
```diff
@@ -0,0 +1,92 @@
+  number: 0
+  skip: True  # [py<3.6]
+
+requirements:
```
It looks like there are too many dependencies in each requirements sub-section.

The build section is for build tools like compilers, cmake, etc. These packages must match the host platform when cross compiling.

The host section is for compiled libraries and python build requirements like python, setuptools, libpython, sometimes numpy (if compiled against their c API).

The run section is for packages only needed at runtime like scipy, matplotlib. Cython is rarely needed here.

Packages only needed to run the tests, go in a separate test/requires section. i.e. pytest

stoiver replied:
@carterbox  thanks. I have tried to minimize the requirements as suggested. Just pushed changes so hopefully I have left all the requirements.

carterbox commented on recipes/anuga/meta.yaml:
```diff
@@ -0,0 +1,92 @@
+    - pip
+  run:
+    - python
+    - m2w64-gcc-libs  # [win]
+    - libpython       # [win]
+    - libstdcxx-ng    # [linux]
+    - libgcc-ng       # [linux]
```
```suggestion
```
These are added automatically as run_exports from the compilers, and many times the cxx standard library isn't even used.

stoiver replied:
@carterbox  removed as suggested.

carterbox commented on recipes/anuga/meta.yaml:
```diff
@@ -0,0 +1,92 @@
+    - utm
+    - wheel
+    - pip
+  run:
```
These run dependencies should have version specifications that match the metadata in the setup.py.

stoiver replied:
@carterbox I ended up removing these from host requirements.

carterbox replied:
The packages in requirements/run should have version specifications which match upstream.

https://github.com/anuga-community/anuga_core/blob/c657e1e41900bcf47f568b2b9bf75892bb46430a/setup.py#L51-L71

carterbox reviewed (changes requested):
This package is using an uncommon layout and build system. Thanks for you patience in advance.

stoiver commented:
@conda-forge/help-python-c This recipe should be ready for review.

carterbox commented on recipes/anuga/meta.yaml:
```diff
@@ -0,0 +1,66 @@
+    - matplotlib-base
+    - meshpy
+    - netcdf4
+    - numpy
```
```suggestion
    - {{ pin_compatible('numpy') }}
```
The numpy version at compile time should match the numpy version at runtime.

carterbox replied:
Just want to double check: Does your compiled code link against the numpy C API?

conda-forge-linter commented:
Hi! This is the friendly automated conda-forge-linting service.

I wanted to let you know that I linted all conda-recipes in your PR (```recipes/anuga```) and found some lint.

Here's what I've got...


For **recipes/anuga**:

 * ``requirements: run: gdal >= 3.0.4`` should not contain a space between relational operator and the version, i.e. ``gdal >=3.0.4``

conda-forge-linter commented:
Hi! This is the friendly automated conda-forge-linting service.

I just wanted to let you know that I linted all conda-recipes in your PR (```recipes/anuga```) and found it was in an excellent condition.

stoiver commented:
@conda-forge/help-python-c This recipe should be ready for review.

@carterbox, I changed the numpy requirement to {{ pin_compatible('numpy') }} as suggested. 

I actually went back to the setup.py include requirements and removed most of the specific version constraints (leaving just a constraint on gdal >=3.0.4).

carterbox commented on recipes/anuga/meta.yaml:
```diff
@@ -0,0 +1,66 @@
+  skip: True  # [py<3.6]
+
+requirements:
+  build:
+    - {{ compiler('c') }}        # [not win]
+    - {{ compiler('cxx') }}      # [not win]
+    - m2w64-gcc                  # [win]
+    - libpython                  # [win]
+
+  host:
+    - python 
+    - libpython  # [win]
+    - pip
+    - cython
+    - numpy
+
+  run:
+    - python
+    - _openmp_mutex   # [linux]
```
```suggestion
  build:
    - {{ compiler('c') }}        # [not win]
    - {{ compiler('cxx') }}      # [not win]
    - m2w64-gcc                  # [win]
    - libpython                  # [win]
    - llvm-openmp  # [osx]
    - libgomp      # [linux]

  host:
    - python 
    - libpython  # [win]
    - pip
    - cython
    - numpy

  run:
    - python
```
https://conda-forge.org/docs/maintainer/knowledge_base.html#openmp

carterbox commented on recipes/anuga/meta.yaml:
```diff
@@ -0,0 +1,66 @@
+    - netcdf4
+    - Pmw
+    - pymetis
+    - pytest
```
```suggestion
```
Why does your package require pytest at runtime? pytest is ususally for development only.

stoiver replied:
@carterbox, we are using the test  structure provided by numpy which based on pytest. And in addition we use pytest to skip some of those tests. With pytest we can run the tests via  `import anuga; anuga.test()`. So should keep `pytest` in the run requirements.

carterbox commented on recipes/anuga/meta.yaml:
```diff
@@ -0,0 +1,66 @@
+    - dill
+    - future
+    - gdal >=3.0.4
+    - gitpython
```
```suggestion
```
gitpython also seems like development tool that may not be needed for the main functionality of modeling floods.

stoiver commented:
@conda-forge/help-python-c This recipe should be ready for review.

@carterbox, regarding "Does your compiled code link against the numpy C API?" We use cython to link between our low level C (and in one case C++) code and python. I don't think we now have any explicit use of the numpy C API, other than through the C code created by cython.
